client/hwlib/src/collectors/hardware_info.rs

```diff
@@ -0,0 +1,338 @@
+/* Copyright 2024 Canonical Ltd.
+ * All rights reserved.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * Written by:
+ *        Nadzeya Hutsko <nadzeya.hutsko@canonical.com>
+ */
+
+use smbioslib;
+use time::macros::format_description;
+use time::Date;
+
+use crate::constants::PROC_DEVICE_TREE_DIRPATH;
+use crate::models::devices;
+
+pub fn load_smbios_data(
+    entry_filepath: Option<&'static str>,
+    table_filepath: Option<&'static str>,
+) -> Result<Option<smbioslib::SMBiosData>, Box<dyn std::error::Error>> {
+    let entry_file = entry_filepath.unwrap_or(smbioslib::SYS_ENTRY_FILE);
+    let table_file = table_filepath.unwrap_or(smbioslib::SYS_TABLE_FILE);
+
+    let smbios_data = match table_load_from_device(entry_file, table_file) {
+        Ok(data) => Some(data),
+        Err(e) => {
+            eprintln!("Failed to load SMBIOS data: {}.", e);
+            None
+        }
+    };
+    Ok(smbios_data)
+}
+
+pub fn collect_bios_info(
+    smbios_data: &smbioslib::SMBiosData,
+) -> Result<devices::Bios, Box<dyn std::error::Error>> {
+    let bios_info = &smbios_data.collect::<smbioslib::SMBiosInformation>()[0];
```

Comment:
Are you guaranteed that this collection (I'm assuming a Vec?) will never be empty? Indexing in will panic if there are no elements while [get](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.get) will give you an option that you can check.

Reply:
Fixed 9f14753